aws/data_source_aws_region.go

```diff
@@ -36,48 +35,74 @@ func dataSourceAwsRegion() *schema.Resource {
 }
 
 func dataSourceAwsRegionRead(d *schema.ResourceData, meta interface{}) error {
-	conn := meta.(*AWSClient).ec2conn
 	currentRegion := meta.(*AWSClient).region
 
-	req := &ec2.DescribeRegionsInput{}
+	var matchingRegion *endpoints.Region
 
-	req.RegionNames = make([]*string, 0, 2)
-	if name := d.Get("name").(string); name != "" {
-		req.RegionNames = append(req.RegionNames, aws.String(name))
+	if v, ok := d.GetOk("endpoint"); ok {
+		endpoint := v.(string)
+		for _, partition := range endpoints.DefaultPartitions() {
+			for _, region := range partition.Regions() {
+				regionEndpointEc2, err := region.ResolveEndpoint(endpoints.Ec2ServiceID)
+				if err != nil {
+					return err
+				}
+				if strings.TrimPrefix(regionEndpointEc2.URL, "https://") == endpoint {
+					matchingRegion = &region
+					break
+				}
+			}
+		}
+		if matchingRegion == nil {
+			return fmt.Errorf("region not found for endpoint: %s", endpoint)
+		}
 	}
 
-	if d.Get("current").(bool) {
-		req.RegionNames = append(req.RegionNames, aws.String(currentRegion))
+	if v, ok := d.GetOk("name"); ok {
+		name := v.(string)
+		for _, partition := range endpoints.DefaultPartitions() {
+			for _, region := range partition.Regions() {
+				if region.ID() == name {
+					if matchingRegion != nil && (*matchingRegion).ID() != name {
+						return fmt.Errorf("multiple regions matched; use additional constraints to reduce matches to a single region")
+					}
+					matchingRegion = &region
+					break
+				}
+			}
+		}
+		if matchingRegion == nil {
+			return fmt.Errorf("region not found for name: %s", name)
+		}
 	}
 
-	req.Filters = buildEC2AttributeFilterList(
-		map[string]string{
-			"endpoint": d.Get("endpoint").(string),
-		},
-	)
-	if len(req.Filters) == 0 {
-		// Don't send an empty filters list; the EC2 API won't accept it.
-		req.Filters = nil
+	current := d.Get("current").(bool)
+	for _, partition := range endpoints.DefaultPartitions() {
+		for _, region := range partition.Regions() {
+			if region.ID() == currentRegion {
+				if matchingRegion == nil {
+					matchingRegion = &region
+					break
+				}
+				if current && (*matchingRegion).ID() != currentRegion {
+					return fmt.Errorf("multiple regions matched; use additional constraints to reduce matches to a single region")
```

Comment:
Is there any reason we can't just error out here regardless of `current`?

Reply:
Ha, I think I was trying to make it so it could support current being false or something. I'm on board with deprecating the `current` attribute completely (making it have no effect in the interim period) so this extra logic is moot and I'll remove it now.